File: solve-x-main/backend/models/student.go
package models

import (
	"errors"
	"time"

	jwt_service "github.com/OmarDardery/solve-the-x-backend/jwt_service"
	mail_service "github.com/OmarDardery/solve-the-x-backend/mail_service"
	"gorm.io/gorm"
)

type Student struct {
	gorm.Model
	FirstName           string
	LastName            string
	Email               string `gorm:"unique"`
	Password            string
	LastChangedPassword time.Time
	Tags                []Tag `gorm:"many2many:student_tags;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	Coins               Coins `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
}

// Generate JWT for the student
func (s *Student) GetJWT() (string, error) {
	return jwt_service.GenerateJWT(s.ID, s.Email, "student")
}

// AuthenticateStudent checks credentials and returns the student if valid
func AuthenticateStudent(db *gorm.DB, email, password string) (*Student, error) {
	var student Student

	if err := db.Where("email = ?", email).First(&student).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("student not found")
		}
		return nil, err
	}

	if !CheckPasswordHash(password, student.Password) {
		return nil, errors.New("invalid credentials")
	}

	return &student, nil
}

// CreateStudent registers a new student and automatically creates a Coins record
func CreateStudent(db *gorm.DB, firstName, lastName, email, password string) error {
	var existing Student
	if err := db.Where("email = ?", email).First(&existing).Error; err == nil {
		return errors.New("email already registered")
	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
		return err
	}

	passwordHash, err := HashPassword(password)
	if err != nil {
		return err
	}

	student := &Student{
		FirstName:           firstName,
		LastName:            lastName,
		Email:               email,
		Password:            passwordHash,
		LastChangedPassword: time.Now(),
	}

	return db.Transaction(func(tx *gorm.DB) error {
		if err := tx.Create(student).Error; err != nil {
			return err
		}

		if err := CreateCoins(tx, student.ID); err != nil {
			return err
		}

		return nil
	})
}

func (s Student) Notify(mailman *mail_service.Mailman, subject, content string) error {
	if mailman == nil {
		// Fallback to legacy function for backward compatibility
		return mail_service.SendNotification(s.Email, subject, content)
	}
	return mailman.SendNotification(s.Email, subject, content)
}

// GetStudentByID retrieves a student by ID with related data (Tags, Coins)
func GetStudentByID(db *gorm.DB, id uint) (*Student, error) {
	var student Student
	if err := db.Preload("Tags").Preload("Coins").First(&student, id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("student not found")
		}
		return nil, err
	}
	return &student, nil
}

// UpdateStudent updates a student’s fields
// UpdateStudent updates a student’s fields and rehashes password if changed
func UpdateStudent(db *gorm.DB, id uint, updates map[string]interface{}) (*Student, error) {
	student, err := GetStudentByID(db, id)
	if err != nil {
		return nil, err
	}

	// Check if password is being updated
	if newPasswordRaw, ok := updates["Password"]; ok {
		newPassword, ok := newPasswordRaw.(string)
		if !ok {
			return nil, errors.New("password must be a string")
		}

		hashed, err := HashPassword(newPassword)
		if err != nil {
			return nil, err
		}

		updates["Password"] = hashed
		updates["LastChangedPassword"] = time.Now()
	}

	if err := db.Model(student).Updates(updates).Error; err != nil {
		return nil, err
	}

	// Reload with relations
	if err := db.Preload("Tags").Preload("Coins").First(student, id).Error; err != nil {
		return nil, err
	}

	return student, nil
}

// DeleteStudent deletes a student by ID
func DeleteStudent(db *gorm.DB, id uint) error {
	student, err := GetStudentByID(db, id)
	if err != nil {
		return err
	}

	if err := db.Delete(student).Error; err != nil {
		return err
	}

	return nil
}

// StudentEmailExists checks if a student with the given email exists
func StudentEmailExists(db *gorm.DB, email string) bool {
	var student Student
	err := db.Where("email = ?", email).First(&student).Error
	return err == nil
}

File: lib/services/student.ts
import prisma from "@/lib/prisma";
import { hashPassword, checkPasswordHash, generateJWT } from "./auth";

/**
 * Create a new student with associated coins record
 */
export async function createStudent(
  firstName: string,
  lastName: string,
  email: string,
  password: string
) {
  // Check if email already exists
  const existing = await prisma.student.findUnique({ where: { email } });
  if (existing) {
    throw new Error("Email already registered");
  }

  const passwordHash = await hashPassword(password);

  // Create student with coins in a transaction
  const student = await prisma.$transaction(async (tx) => {
    const newStudent = await tx.student.create({
      data: {
        firstName,
        lastName,
        email,
        password: passwordHash,
        lastChangedPassword: new Date(),
      },
    });

    // Create coins record for student
    await tx.coins.create({
      data: {
        amount: 0,
        studentId: newStudent.id,
      },
    });

    return newStudent;
  });

  return student;
}

/**
 * Authenticate a student and return the student if valid
 */
export async function authenticateStudent(email: string, password: string) {
  const student = await prisma.student.findUnique({ where: { email } });
  
  if (!student) {
    throw new Error("Student not found");
  }

  if (!student.password) {
    throw new Error("Invalid credentials");
  }

  const isValid = await checkPasswordHash(password, student.password);
  if (!isValid) {
    throw new Error("Invalid credentials");
  }

  return student;
}

/**
 * Get JWT for student
 */
export function getStudentJWT(student: { id: bigint; email: string | null }) {
  return generateJWT(student.id, student.email || "", "student");
}

/**
 * Get student by ID with related data
 */
export async function getStudentById(id: bigint) {
  const student = await prisma.student.findUnique({
    where: { id },
    include: {
      student_tags: {
        include: { tags: true },
      },
      coins: true,
    },
  });

  if (!student) {
    throw new Error("Student not found");
  }

  return student;
}

/**
 * Update student fields
 */
export async function updateStudent(
  id: bigint,
  updates: {
    firstName?: string;
    lastName?: string;
    password?: string;
    [key: string]: any;
  }
) {
  // If password is being updated, hash it
  if (updates.password) {
    updates.password = await hashPassword(updates.password);
    updates.lastChangedPassword = new Date();
  }

  const student = await prisma.student.update({
    where: { id },
    data: updates,
    include: {
      student_tags: {
        include: { tags: true },
      },
      coins: true,
    },
  });

  return student;
}

/**
 * Delete a student
 */
export async function deleteStudent(id: bigint) {
  await prisma.student.delete({ where: { id } });
}

/**
 * Check if student email exists
 */
export async function studentEmailExists(email: string): Promise<boolean> {
  const student = await prisma.student.findUnique({ where: { email } });
  return !!student;
}
